harmony/harmonytask/task_type_handler.go

```diff
@@ -63,45 +66,151 @@ const (
 // The only caller should be the one work poller thread. This does spin off other threads,
 // but those should not considerWork. Work completing may lower the resource numbers
 // unexpectedly, but that will not invalidate work being already able to fit.
-func (h *taskTypeHandler) considerWork(from string, ids []TaskID) (workAccepted bool) {
-top:
+func (h *taskTypeHandler) considerWork(from string, ids []TaskID) (workAccepted, liveBids bool) {
 	if len(ids) == 0 {
-		return true // stop looking for takers
+		return true, false // stop looking for takers
 	}
 
 	// 1. Can we do any more of this task type?
 	// NOTE: 0 is the default value, so this way people don't need to worry about
 	// this setting unless they want to limit the number of tasks of this type.
 	if h.Max > 0 && int(h.Count.Load()) >= h.Max {
 		log.Debugw("did not accept task", "name", h.Name, "reason", "at max already")
-		return false
+		return false, false
 	}
 
 	// 2. Can we do any more work? From here onward, we presume the resource
 	// story will not change, so single-threaded calling is best.
 	err := h.AssertMachineHasCapacity()
 	if err != nil {
 		log.Debugw("did not accept task", "name", h.Name, "reason", "at capacity already: "+err.Error())
-		return false
+		return false, false
 	}
 
-	h.TaskEngine.WorkOrigin = from
-
-	// 3. What does the impl say?
-canAcceptAgain:
-	tID, err := h.CanAccept(ids, h.TaskEngine)
+	return h.canAcceptPart(from, ids)
+}
+func (h *taskTypeHandler) canAcceptPart(from string, ids []TaskID) (workAccepted, liveBids bool) {
+	sch := &SchedulingInfo{h.TaskEngine, from}
 
-	h.TaskEngine.WorkOrigin = ""
+	// 3. Bid?
+	if h.Bid {
+		return h.BidLogic(from, ids, sch)
+	}
 
+	// 3b. Or run directly?
+	tID, err := h.TaskInterface.(FastTask).CanAccept(ids, sch)
 	if err != nil {
 		log.Error(err)
-		return false
+		return false, false
 	}
 	if tID == nil {
 		log.Infow("did not accept task", "task_id", ids[0], "reason", "CanAccept() refused", "name", h.Name)
-		return false
+		return false, false
+	}
+	return h.startTask(from, tID, ids), false
+}
+
+func (h *taskTypeHandler) GetUnownedTasks() (ids []TaskID, err error) {
+	var numBidders int
+	err = h.TaskEngine.db.QueryRow(h.TaskEngine.ctx, `SELECT COUNT(*) FROM harmony_machine_details WHERE tasks LIKE $1`, "%,"+h.TaskTypeDetails.Name+",%").Scan(&numBidders)
+	if err != nil {
+		log.Error("Unable to read harmony_machine_details ", err)
+		return nil, err
+	}
+	var unownedTasks []TaskID
+	err = h.TaskEngine.db.Select(h.TaskEngine.ctx, &unownedTasks, `SELECT id
+		FROM harmony_task
+		WHERE owner_id IS NULL AND name=$1
+		ORDER BY creation_time ASC
+		LIMIT $2`, h.TaskTypeDetails.Name, numBidders)
```

Comment:
I don't understand why we are limiting to the number of bidders here. Wouldn't this break everyone should bid on all possible tasks?

Reply:
I'm handling the case when 10000 of something drops in to the queue. 
Do we spend minutes in bidding just to rebid? 
    That could break the healthy uptake in jobs and the bidding timers. 
Do we cache our bids?
  That limits the usefulness of bidding as it cannot include capacity.
So the remaining option is to universally limit the max number of items we care to bid on. 
  - A limit of 1 could be too low for big networks. 
  - A limit of "everyone's capacity for this task" is closer to right, but requires (slow) per-cycle pre-negotiation.
  - A limit of "1x bidders" will keep work flowing, but may take multiple bid rounds to fill bigger machines. 
A future adjustment could include more than 1x per machine if the "handful of behemoths" clusters suffer.